### ISSUE TITLE
Adding MaxNativeZoom option for deeper zooming

### ISSUE THREAD

**Issue by gotjoshua** [mapbox/geojson.io#475]
How would it be possible to add the MaxNativeZoom property to the tile layer to avoid the black X's when i zoom in on the satellite map?
thanks,
joshua

**Comment by tmcw** [comment]
The satellite map has variable zoom levels throughout the globe - most cities are covered until the very end, whereas lower-density places have less zoom level coverage. Restricting the world to the lowest zoom would cut off cities, and there's not yet an API for determining zoom level coverage, so this is not actionable.

**Comment by gotjoshua** [comment]
Maybe I don't understand fully. I am referring to the maxNativeZoom prop in Leaflet - which allows the tile to be scaled (and appear pixelated) rather than displaying black tiles with Xs on them.

http://gis.stackexchange.com/questions/78843/zoom-further-in-with-leaflet-javascirpt-api

In my timemapper fork, i have it working from the simple instructions in the above link.

Is there more complexity in the mapbox system that does not allow this feature to be activated?
Could you explain a bit more for me? 
thanks...

**Comment by tmcw** [comment]
The black x's are images, so they respond with HTTP 200 like other images. Leaflet cannot tell the difference between a black x and any other image, and thus cannot have different scaling rules.

**Comment by gotjoshua** [comment]
Aha! Thanks Tom, that was the clue I needed... I guess that means I can use
a different tileset for the satellite mode (like esri) that won't return
the black xs...
cheers!

~
.:.joshua
"Life is a series of gestures, into which we embed meaning..."

"The fear of being free makes us proud to be slaves."

"Morality is doing what is right regardless of what you are told. Obedience
is doing what you are told regardless of what is right"

**Comment by gotjoshua** [comment]
Bummer, that didn't work as easily as I hoped.
I changed the TileLayer source to:
var Esri_WorldImagery = L.tileLayer('http://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}', {
			attribution: 'Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community',
	  		minZoom: 1,
			maxZoom:22,
			maxNativeZoom: 18
		});

but Mapbox is still requesting zoom level 19 tiles (not respecting the maxNativeZoom property)

any ideas why?

which version of leaflet is currently bundled with geojson.io?

thanks for your engagement!

**Comment by luxigo** [comment]
@tmcw @gotjoshua did you solve this or are you working on it ?

I began to look into mapbox.js (that I replaced with mapbox.uncompressed.js). I was able to zoom further using maxNativeZoom, after returning maxNativeZoom from method getMaxZoom and disabling conditionnaly the 'tileunload' event in method _reset bound to the 'viewreset' event fired from method _resetView...

But there is still some issues left in _resetView: with the 'zoomend' (tiles are removed) and 'moveend' events. Something needs to be done on the geojson.io side too.

Also the fact that the tiles for maxZoom are not loaded when the page is initialized with a zoom level > maxZoom

I cannot spend more time on it for now, but I need the feature: geojson.io is far more convenient than qgis :-)
